backend/src/apiserver/server/api_util.go

```diff
@@ -137,14 +160,57 @@ func GetResourceReferenceFromRunInterface(r interface{}) ([]*apiv1beta1.Resource
 	}
 }
 
-// Converts a yaml string into structpb.Struct (v2).
-func yamlStringToProtobufStruct(s string) (*structpb.Struct, error) {
-	st := &structpb.Struct{}
-	err := yaml.Unmarshal([]byte(s), st)
-	if err != nil {
-		return nil, util.Wrap(err, "Failed to convert a yaml string into a protobuf struct")
+// Converts pipeline spec in yaml string into structpb.Struct (v2).
+// The string may contain multiple yaml documents, where there has to be one doc for pipeline spec,
+// and there may be a second doc for platform specific spec.
+// If platform spec is empty, then return pipeline spec directly. Else, return a struct with
+// pipeline spec and platform spec as subfields.
+func yamlStringToPipelineSpecStruct(s string) (*structpb.Struct, error) {
+	if s == "" {
+		return nil, nil
+	}
+	var pipelineSpec structpb.Struct
+	var platformSpec structpb.Struct
+	hasPipelineSpec := false
+	hasPlatformSpec := false
+
+	yamlStrings := strings.Split(s, "\n---\n")
+	for _, yamlString := range yamlStrings {
+		if template.IsPlatformSpecWithKubernetesConfig([]byte(yamlString)) {
+			hasPlatformSpec = true
+			jsonBytes, err := yaml.YAMLToJSON([]byte(yamlString))
+			if err != nil {
+				return nil, util.Wrap(err, "Failed to convert pipelineSpec yaml string into a protobuf struct")
+			}
+			err = protojson.Unmarshal(jsonBytes, &platformSpec)
+			if err != nil {
+				return nil, util.Wrap(err, "Failed to convert platformSpec yaml string into a protobuf struct")
+			}
+		} else {
+			hasPipelineSpec = true
+			jsonBytes, err := yaml.YAMLToJSON([]byte(yamlString))
+			if err != nil {
+				return nil, util.Wrap(err, "Failed to convert pipelineSpec yaml string into a protobuf struct")
+			}
+			err = protojson.Unmarshal(jsonBytes, &pipelineSpec)
+			if err != nil {
+				return nil, util.Wrap(err, "Failed to convert pipelineSpec yaml string into a protobuf struct")
+			}
+		}
+	}
+	if !hasPipelineSpec {
+		return nil, util.NewInvalidInputError("No pipeline spec provided in yaml doc")
+	} else if !hasPlatformSpec {
+		return &pipelineSpec, nil
+	} else {
+		pipeline := &structpb.Struct{
+			Fields: map[string]*structpb.Value{
+				"pipelineSpec": structpb.NewStructValue(&pipelineSpec),
+				"platforms":    structpb.NewStructValue(&platformSpec),
+			},
+		}
+		return pipeline, nil
 	}
```

Comment:
This is probably intended behavior, but I wanted to call this out. For a YAML file like this
```yaml
<pipelineSpec>
---
<empty or invalid (e.g. missing `kubernetes` key) platformSpec>
```
We will ge
```golang
return &pipelineSpec, nil
```
So we intentionally dropping data with `IsPlatformSpecWithKubernetesConfi()` and treat it as the old-style `pipelineSpec`.

Reply:
That's right. With `platformSpec` empty, the backend cannot tell how `pipelineSpec` was passed in.